newlib/libc/sys/vita/socket.c

```diff
@@ -491,3 +493,58 @@ int	socket(int domain, int type, int protocol)
 	return s;
 }
 #endif
+
+#ifdef F_socket
+int	socketpair(int domain, int type, int protocol, int sockfds[2])
+{
+	// Usually socketpair is used with AF_UNIX, simulate that with INET.
+	int listener;
+	if ((listener = socket(AF_INET, type, protocol)) == -1) {
+		return -1;
+	}
+
+	struct sockaddr_in server_addr;
+	socklen_t addr_len = sizeof(struct sockaddr_in);
+
+	memset(&server_addr, 0, sizeof(server_addr));
+	server_addr.sin_family = AF_INET;
+	server_addr.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
+	server_addr.sin_port = 0;
+
+	if (bind(listener, (struct sockaddr*)&server_addr, addr_len) == -1) {
+		close(listener);
+		return -1;
+	}
+
+	if (listen(listener, 1) == -1) {
+		close(listener);
+		return -1;
+	}
+
+	if (getsockname(listener, (struct sockaddr *)&server_addr, &addr_len) == -1) {
+		close(listener);
+		return -1;
+	}
+
+	if ((sockfds[1] = socket(AF_INET, type, protocol)) == -1) {
+		close(listener);
+		return -1;
+	}
+
+	if (connect(sockfds[1], (struct sockaddr*)&server_addr, addr_len) == -1) {
+		close(sockfds[1]);
+		close(listener);
+		return -1;
+	}
+
+	if ((sockfds[0] = accept(listener, (struct sockaddr*)&server_addr, &addr_len)) == -1) {
+		close(sockfds[1]);
+		close(listener);
+		return -1;
+	}
```

Comment:
@nikarh i love this pr.
just one question, don't we need to set [errno](https://pubs.opengroup.org/onlinepubs/9699919799/functions/socketpair.html)?

maybe almost errors already set by these base calls.
if you confirm it, i'll merge it

Reply:
Since this impl delegates to other newlib methods (connect, accept, socket), they already set `errno`, so this shouldn't be a problem